File: benchmarks/executable/restaurant_uint_arithmetic.c
// C23 reference for the Restaurant checked UInt/u64 arithmetic workload.

#include <inttypes.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

typedef struct {
    bool ok;
    uint64_t value;
} checked_u64;

/*
 * Volatile operands keep every arithmetic and comparison operation dependent
 * on a runtime load in optimized builds. The values deliberately include the
 * first unsigned value above INT64_MAX and UINT64_MAX.
 */
static volatile uint64_t runtime_high = UINT64_C(9223372036854775808);
static volatile uint64_t runtime_increment = UINT64_C(2);
static volatile uint64_t runtime_one = UINT64_C(1);
static volatile uint64_t runtime_product_left = UINT64_C(3);
static volatile uint64_t runtime_product_right = UINT64_C(7);
static volatile uint64_t runtime_dividend = UINT64_C(23);
static volatile uint64_t runtime_divisor = UINT64_C(3);
static volatile uint64_t runtime_max = UINT64_MAX;
static volatile uint64_t runtime_signed_max = UINT64_C(9223372036854775807);
static volatile uint64_t runtime_high_equal = UINT64_C(9223372036854775808);

static checked_u64 checked_add_u64(uint64_t left, uint64_t right) {
    if (left > UINT64_MAX - right) {
        return (checked_u64){.ok = false, .value = 0};
    }
    return (checked_u64){.ok = true, .value = left + right};
}

static checked_u64 checked_subtract_u64(uint64_t left, uint64_t right) {
    if (left < right) {
        return (checked_u64){.ok = false, .value = 0};
    }
    return (checked_u64){.ok = true, .value = left - right};
}

static checked_u64 checked_multiply_u64(uint64_t left, uint64_t right) {
    if (left != 0 && right > UINT64_MAX / left) {
        return (checked_u64){.ok = false, .value = 0};
    }
    return (checked_u64){.ok = true, .value = left * right};
}

static checked_u64 checked_divide_u64(uint64_t left, uint64_t right) {
    if (right == 0) {
        return (checked_u64){.ok = false, .value = 0};
    }
    return (checked_u64){.ok = true, .value = left / right};
}

static checked_u64 checked_remainder_u64(uint64_t left, uint64_t right) {
    if (right == 0) {
        return (checked_u64){.ok = false, .value = 0};
    }
    return (checked_u64){.ok = true, .value = left % right};
}

int main(void) {
#ifdef _WIN32
    if (_setmode(_fileno(stdout), _O_BINARY) == -1) return 1;
#endif

    const uint64_t high = runtime_high;
    const uint64_t increment = runtime_increment;
    const uint64_t one = runtime_one;
    const uint64_t product_left = runtime_product_left;
    const uint64_t product_right = runtime_product_right;
    const uint64_t dividend = runtime_dividend;
    const uint64_t divisor = runtime_divisor;
    const uint64_t max = runtime_max;
    const uint64_t signed_max = runtime_signed_max;
    const uint64_t high_equal = runtime_high_equal;

    const checked_u64 sum = checked_add_u64(high, increment);
    const checked_u64 difference = checked_subtract_u64(sum.value, one);
    const checked_u64 product = checked_multiply_u64(product_left, product_right);
    const checked_u64 quotient = checked_divide_u64(dividend, divisor);
    const checked_u64 remainder = checked_remainder_u64(dividend, divisor);
    const bool comparison_high = high > signed_max;
    const bool comparison_high_equal = high >= high_equal;
    const bool comparison_max = max > high;
    const bool comparison_high_less = high < max;
    const bool comparison_signed_equal = high == signed_max;
    const bool comparison_signed_different = high != signed_max;
    const bool comparison_less_equal = high <= high_equal;

    if (!sum.ok || !difference.ok || !product.ok || !quotient.ok ||
        !remainder.ok || !comparison_high || !comparison_high_equal ||
        !comparison_max || !comparison_high_less || comparison_signed_equal ||
        !comparison_signed_different || !comparison_less_equal) {
        return 1;
    }

    if (printf("UInt %" PRIu64 "/%" PRIu64 "/%" PRIu64
               "; div %" PRIu64 "; rem %" PRIu64
               "; cmp %s/%s/%s/%s/%s/%s/%s\n",
               sum.value, difference.value, product.value, quotient.value,
               remainder.value, comparison_high ? "true" : "false",
               comparison_high_equal ? "true" : "false",
               comparison_max ? "true" : "false",
               comparison_high_less ? "true" : "false",
               comparison_signed_equal ? "true" : "false",
               comparison_signed_different ? "true" : "false",
               comparison_less_equal ? "true" : "false") < 0) {
        return 1;
    }
    return 0;
}

File: benchmarks/executable/restaurant_uint_arithmetic.rs
// Rust 2024 reference for the Restaurant checked UInt/u64 arithmetic workload.

use std::hint::black_box;
use std::io::{self, Write};

#[derive(Clone, Copy)]
struct CheckedU64 {
    ok: bool,
    value: u64,
}

fn checked_add_u64(left: u64, right: u64) -> CheckedU64 {
    match left.checked_add(right) {
        Some(value) => CheckedU64 { ok: true, value },
        None => CheckedU64 { ok: false, value: 0 },
    }
}

fn checked_subtract_u64(left: u64, right: u64) -> CheckedU64 {
    match left.checked_sub(right) {
        Some(value) => CheckedU64 { ok: true, value },
        None => CheckedU64 { ok: false, value: 0 },
    }
}

fn checked_multiply_u64(left: u64, right: u64) -> CheckedU64 {
    match left.checked_mul(right) {
        Some(value) => CheckedU64 { ok: true, value },
        None => CheckedU64 { ok: false, value: 0 },
    }
}

fn checked_divide_u64(left: u64, right: u64) -> CheckedU64 {
    match left.checked_div(right) {
        Some(value) => CheckedU64 { ok: true, value },
        None => CheckedU64 { ok: false, value: 0 },
    }
}

fn checked_remainder_u64(left: u64, right: u64) -> CheckedU64 {
    match left.checked_rem(right) {
        Some(value) => CheckedU64 { ok: true, value },
        None => CheckedU64 { ok: false, value: 0 },
    }
}

fn main() {
    // black_box keeps each reference operand runtime-dependent in Release.
    let high = black_box(9_223_372_036_854_775_808_u64);
    let increment = black_box(2_u64);
    let one = black_box(1_u64);
    let product_left = black_box(3_u64);
    let product_right = black_box(7_u64);
    let dividend = black_box(23_u64);
    let divisor = black_box(3_u64);
    let max = black_box(u64::MAX);
    let signed_max = black_box(i64::MAX as u64);
    let high_equal = black_box(9_223_372_036_854_775_808_u64);

    let sum = checked_add_u64(high, increment);
    let difference = checked_subtract_u64(sum.value, one);
    let product = checked_multiply_u64(product_left, product_right);
    let quotient = checked_divide_u64(dividend, divisor);
    let remainder = checked_remainder_u64(dividend, divisor);
    let comparison_high = high > signed_max;
    let comparison_high_equal = high >= high_equal;
    let comparison_max = max > high;
    let comparison_high_less = high < max;
    let comparison_signed_equal = high == signed_max;
    let comparison_signed_different = high != signed_max;
    let comparison_less_equal = high <= high_equal;

    if !sum.ok ||
        !difference.ok ||
        !product.ok ||
        !quotient.ok ||
        !remainder.ok ||
        !comparison_high ||
        !comparison_high_equal ||
        !comparison_max ||
        !comparison_high_less ||
        comparison_signed_equal ||
        !comparison_signed_different ||
        !comparison_less_equal
    {
        std::process::exit(1);
    }

    let mut stdout = io::stdout().lock();
    if writeln!(
        stdout,
        "UInt {}/{}/{}; div {}; rem {}; cmp {}/{}/{}/{}/{}/{}/{}",
        sum.value,
        difference.value,
        product.value,
        quotient.value,
        remainder.value,
        comparison_high,
        comparison_high_equal,
        comparison_max,
        comparison_high_less,
        comparison_signed_equal,
        comparison_signed_different,
        comparison_less_equal,
    )
    .is_err()
    {
        std::process::exit(1);
    }
}
